controllers/component_controller_test.go

```diff
@@ -0,0 +1,562 @@
+
+			// project should be set in hasApp
+			hasProjects, err := hasAppDevfile.GetProjects(common.DevfileOptions{})
+			Expect(err).Should(Not(HaveOccurred()))
+			Expect(len(hasProjects)).ShouldNot(Equal(0))
+
+			nameMatched := false
+			repoLinkMatched := false
+			for _, project := range hasProjects {
+				if project.Name == ComponentName {
+					nameMatched = true
+				}
+				if project.Git != nil && project.Git.GitLikeProjectSource.Remotes["origin"] == SampleRepoLink {
+					repoLinkMatched = true
+				}
+			}
+			Expect(nameMatched).Should(Equal(true))
+			Expect(repoLinkMatched).Should(Equal(true))
+
+			// update the hasComp and apply
+			createdHasComp.Spec.Replicas = updatedReplica
+			createdHasComp.Spec.Route = updatedRoute
+			createdHasComp.Spec.TargetPort = updatedPort
+			createdHasComp.Spec.Env = updatedEnv
+			createdHasComp.Spec.Resources = updatedResources
+			Expect(k8sClient.Update(ctx, createdHasComp)).Should(Succeed())
+
+			updatedHasComp := &appstudiov1alpha1.Component{}
+			Eventually(func() bool {
+				k8sClient.Get(context.Background(), hasCompLookupKey, updatedHasComp)
+				return updatedHasComp.Status.Conditions[len(updatedHasComp.Status.Conditions)-1].Type == "Updated"
+			}, timeout, interval).Should(BeTrue())
+
+			// Make sure the devfile model was properly set in Component
+			Expect(updatedHasComp.Status.Devfile).Should(Not(Equal("")))
+
+			// Check the HAS Component updated devfile
+			hasCompUpdatedDevfile, err := devfile.ParseDevfileModel(updatedHasComp.Status.Devfile)
+			Expect(err).Should(Not(HaveOccurred()))
+
+			checklist = updateChecklist{
+				route:     updatedRoute,
+				replica:   updatedReplica,
+				port:      updatedPort,
+				env:       updatedEnv,
+				resources: updatedResources,
+			}
+
+			verifyHASComponentUpdates(hasCompUpdatedDevfile, checklist, nil)
+
+			// Delete the specified HASComp resource
+			deleteHASCompCR(hasCompLookupKey)
+
+			// Delete the specified HASApp resource
+			deleteHASAppCR(hasAppLookupKey)
+		})
+	})
+
+})
+
+type updateChecklist struct {
+	route     string
+	port      int
+	replica   int
+	env       []corev1.EnvVar
+	resources corev1.ResourceRequirements
+}
+
+func verifyHASComponentUpdates(devfile data.DevfileData, checklist updateChecklist, goPkgTest *testing.T) {
+	// container component should be updated with the necessary hasComp properties
+	components, err := devfile.GetComponents(common.DevfileOptions{
+		ComponentOptions: common.ComponentOptions{
+			ComponentType: v1alpha2.ContainerComponentType,
+		},
+	})
+	if goPkgTest == nil {
+		Expect(err).Should(Not(HaveOccurred()))
+	} else if err != nil {
+		goPkgTest.Error(err)
+	}
+
+	requests := checklist.resources.Requests
+	limits := checklist.resources.Limits
+
+	for i, component := range components {
+		attributes := component.Attributes
+		var err error
+
+		// Check the route
+		route := attributes.Get("appstudio.has/route", &err)
```

Comment:
Not sure if I'm following along fully: What happens `appstudio.has/route` (and other attributes) aren't set in the devfile for a valid test case?

Reply:
yeah currently I am setting it everytime I call this func, let me see what i can do

Reply:
This test func now checks if the checklist arguments are present before checking